File: workflow/scripts/rcorrector_cleanup.py
import json
import gzip
import sys
from contextlib import ExitStack

metric_string = " l:"
corrected_string = " cor\n"
unfixed_string = " unfixable_error\n"

data = json.loads(sys.argv[1])

with ExitStack() as stack:
    fwd_input = stack.enter_context(gzip.open(data["input"][0], "rt"))
    rev_input = stack.enter_context(gzip.open(data["input"][1], "rt"))
    fwd_output = stack.enter_context(gzip.open(data["output"][0], "wt"))
    rev_output = stack.enter_context(gzip.open(data["output"][1], "wt"))
    logger = stack.enter_context(open(data["log"], "w"))
    
    pair_total_count = 0
    pair_corrected_count = 0
    pair_removed_count = 0
    
    while True:
        fwd_read = fwd_input.readline()
        rev_read = rev_input.readline()
        
        if fwd_read == "" or rev_read == "":
            break
        
        pair_total_count += 1
        
        if fwd_read.endswith(unfixed_string) or rev_read.endswith(unfixed_string):
            pair_removed_count += 1
            
            fwd_input.readline()
            fwd_input.readline()
            fwd_input.readline()
            rev_input.readline()
            rev_input.readline()
            rev_input.readline()
            
            continue
        
        if fwd_read.endswith(corrected_string) or rev_read.endswith(corrected_string):
            pair_corrected_count += 1
        
        fwd_output.write(fwd_read[:fwd_read.rfind(metric_string)] + "\n")
        rev_output.write(rev_read[:fwd_read.rfind(metric_string)] + "\n")
        
        fwd_output.write(fwd_input.readline())
        fwd_output.write(fwd_input.readline())
        fwd_output.write(fwd_input.readline())
        rev_output.write(rev_input.readline())              
        rev_output.write(rev_input.readline())
        rev_output.write(rev_input.readline())
    
    logger.write(f"Input forward reads: {data['input'][0]}\n")
    logger.write(f"Input reverse reads: {data['input'][1]}\n")
    logger.write(f"Output forward reads: {data['output'][0]}\n")
    logger.write(f"Output reverse reads: {data['output'][1]}\n")
    logger.write(f"Total read pairs processed: {pair_total_count}\n")
    logger.write(f"Total read pairs corrected: {pair_corrected_count}\n")
    logger.write(f"Total read pairs removed: {pair_removed_count}\n")

File: workflow/scripts/rcorrector_cleanup.rs
//! ```cargo
//! [dependencies]
//! bio = "*"
//! flate2 = { version = "*", features = ["zlib-ng-compat"], default-features = false }
//! ```

use anyhow::Result;
use bio::io::fastq;
use flate2::Compression;
use flate2::read::MultiGzDecoder;
use flate2::write::GzEncoder;
use std::fs::File;
use std::io::{BufReader, Write};
use std::path::Path;

const CORRECTED_STRING: &str = "cor";
const METRIC_STRING: &str  = " l:";
const UNFIXABLE_STRING: &str = "unfixable_error";

fn get_fastq_reader<P: AsRef<Path>>(path: P) -> Result<fastq::Reader<BufReader<MultiGzDecoder<File>>>> {
    let file = File::open(path)?;
    Ok(fastq::Reader::new(MultiGzDecoder::new(file)))
}

fn get_fastq_writer<P: AsRef<Path>>(path: P) -> Result<fastq::Writer<GzEncoder<File>>> {
    let file = File::create(path)?;
    Ok(fastq::Writer::new(GzEncoder::new(file, Compression::fast())))
}

fn get_logger<P: AsRef<Path>>(path: P) -> Result<File> {
    let file = File::create(path)?;
    Ok(file)
}

fn main() -> Result<()> {
    let mut forward_records = get_fastq_reader(snakemake.input[0])?.records();
    let mut reverse_records = get_fastq_reader(snakemake.input[1])?.records();
    let mut output_forward = get_fastq_writer(snakemake.output[0])?;
    let mut output_reverse = get_fastq_writer(snakemake.output[1])?;
    let mut logger = get_logger(snakemake.log[0])?;
    
    let mut total_count = 0;
    let mut corrected_count = 0;
    let mut removed_count = 0;
    
    while let (Some(forward_record), Some(reverse_record)) = (forward_records.next(), reverse_records.next()) {
        let forward_record = forward_record?;
        let reverse_record = reverse_record?;
        
        total_count += 1;
        
        if forward_record.desc().unwrap().ends_with(UNFIXABLE_STRING) || reverse_record.desc().unwrap().ends_with(UNFIXABLE_STRING) {
            removed_count += 1;
            continue;
        }
        
        if forward_record.desc().unwrap().ends_with(CORRECTED_STRING) || reverse_record.desc().unwrap().ends_with(CORRECTED_STRING) {
            corrected_count += 1;
        }
        
        let mut forward_header = String::from(forward_record.id());
        forward_header.push(' ');
        forward_header.push_str(forward_record.desc().unwrap());
        
        let forward_header = forward_header.rsplit_once(METRIC_STRING).unwrap().0;
        
        let mut reverse_header = String::from(reverse_record.id());
        reverse_header.push(' ');
        reverse_header.push_str(reverse_record.desc().unwrap());
        
        let reverse_header = reverse_header.rsplit_once(METRIC_STRING).unwrap().0;
        
        output_forward.write(forward_header, None, forward_record.seq(), forward_record.qual())?;
        output_reverse.write(reverse_header, None, reverse_record.seq(), reverse_record.qual())?;
    }
    
    writeln!(logger, "Input forward reads: {}", snakemake.input[0])?;
    writeln!(logger, "Input reverse reads: {}", snakemake.input[1])?;
    writeln!(logger, "Output forward reads: {}", snakemake.output[0])?;
    writeln!(logger, "Output reverse reads: {}", snakemake.output[1])?;
    writeln!(logger, "Total read pairs processed: {}", total_count)?;
    writeln!(logger, "Total read pairs corrected: {}", corrected_count)?;
    writeln!(logger, "Total read pairs removed: {}", removed_count)?;
    
    output_forward.flush()?;
    output_reverse.flush()?;
    logger.flush()?;
    
    Ok(())
}
